File: slices/scopes/scopes-rust/src/scope/known.rs
//! Broadly-known non-resource scopes (OIDC, launch context, refresh).

/// A broadly-known non-resource scope that matches exactly.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum KnownScope {
    OfflineAccess,
    Openid,
    Profile,
    Launch,
    LaunchPatient,
    FhirUser,
    /// `wildflower/launch` — the umbrella app-launch capability. Holding it
    /// authorizes launching apps at all; a SMART app additionally requires the
    /// caller's grant to cover that app's own requested scopes. It is a *known*
    /// scope, matched exactly, so the `wildflower/*` resource wildcard does **not**
    /// cover it — launch is an explicitly-granted capability even for an owner
    /// token. (The `wildflower/` prefix is not a resource scope: with no `.perms`
    /// segment it never parses as [`WildflowerResourceScope`](super::WildflowerResourceScope).)
    AnyScopedAppLaunch,
}

impl KnownScope {
    pub(in crate::scope) fn parse(s: &str) -> Option<Self> {
        match s {
            "offline_access" => Some(KnownScope::OfflineAccess),
            "openid" => Some(KnownScope::Openid),
            "profile" => Some(KnownScope::Profile),
            "launch" => Some(KnownScope::Launch),
            "launch/patient" => Some(KnownScope::LaunchPatient),
            "fhirUser" => Some(KnownScope::FhirUser),
            "wildflower/launch" => Some(KnownScope::AnyScopedAppLaunch),
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            KnownScope::OfflineAccess => "offline_access",
            KnownScope::Openid => "openid",
            KnownScope::Profile => "profile",
            KnownScope::Launch => "launch",
            KnownScope::LaunchPatient => "launch/patient",
            KnownScope::FhirUser => "fhirUser",
            KnownScope::AnyScopedAppLaunch => "wildflower/launch",
        }
    }
}

File: slices/scopes/scopes-core/src/domain/known.ts
/**
 * Broadly-known non-resource (flag) scopes — `openid`, `offline_access`, the
 * launch-context scopes — mirroring `scopes-rust`'s `KnownScope`
 * (`scope/known.rs`). These match exactly; they are never resource scopes.
 *
 * Namespace module (`import { KnownScope } from 'scopes-core'`): the enum is
 * {@link KnownScope}, with `KnownScope.scopeParse`, `KnownScope.is`, `KnownScope.ALL`.
 */

import type { ScopeParser } from '../behaviour/index.ts'

/** A broadly-known non-resource scope that matches exactly (Rust's `KnownScope`). */
type KnownScope = 'openid' | 'profile' | 'fhirUser' | 'offline_access' | 'launch' | 'launch/patient'

/** The canonical flag scopes in display order (`spec.md §7`). */
const ALL: readonly KnownScope[] = [
  'openid',
  'profile',
  'fhirUser',
  'offline_access',
  'launch',
  'launch/patient',
]

/** Whether a string is a known flag scope. */
const is = (s: string): s is KnownScope => (ALL as readonly string[]).includes(s)

/** Parse a known flag scope, or `null` if `s` isn't one. */
const scopeParse: ScopeParser<KnownScope>['scopeParse'] = (s) => (is(s) ? s : null)

export { type KnownScope, ALL, is, scopeParse }
